AndelaOSP/shufflebox-frontend
# #140317487 Create subviews with routing.

#### What does this PR do?
- Refactors the project setup to allow for reuse of common UI components.
- Improves on the existing client-side routing setup by converting it to use hashless urls.
- Implements reverse routing and removes hardcoded urls.

#### Description of Task to be completed?
Home, About and FAQ pages share common UI components. Hangouts, Secret Santa and Brownbag pages share common UI elements as well. The task is to refactor the code to allow for reuse of the common UI elements by creating subviews instead of including them in every view.

#### How should this be manually tested?
Spin up the server and click around. Notice the URLs don't have hashes. Look through the code and realise that navigating to a new URL only changes a piece of the view, not the whole view. Also notice that we've moved away from hardcoded URLs in favour of reverse routing.

#### Any background context you want to provide?
@akabiru  and I were deliberating on the best way to implement the subviews, seeing as there's no easy way to identify what subview a particular message should invoke a response from. I've ended up having to pattern match twice, once in the main view and again in the subview. This PR, however, lays groundwork for easy refactoring once the authentication functionality is implemented. I propose the duplicated pattern matching be removed then.

**PS**: For reviewers, I'm leaving multiple commits in this PR to make it easier to follow how the proposed changes evolved the application to the current state. The commits should probably be squashed if and when merging this PR.

#### What are the relevant pivotal tracker stories?
[Create sub-views with routing](https://www.pivotaltracker.com/story/show/140317487)
#### Screenshots (if appropriate)
#### Questions:
